CSS nesting: parsing of style rule inside style rule.

mdubet opened the pull request:
#### 2dc309f4a725d902f5c24909e84d7cdabb499c9f
<pre>
CSS nesting: parsing of style rule inside style rule.
<a href="https://bugs.webkit.org/show_bug.cgi?id=249263">https://bugs.webkit.org/show_bug.cgi?id=249263</a>
rdar://103146832

Reviewed by Antti Koivisto.

This implements experimental support to parse nested style rule inside style rule.
This is based on the css-nesting-1 editor draft as of today,
which describes the so-called &quot;option 3&quot; (non letter start) syntax.

<a href="https://github.com/w3c/csswg-drafts/blob/main/css-nesting-1/proposals.md">https://github.com/w3c/csswg-drafts/blob/main/css-nesting-1/proposals.md</a>

This is missing:
  - any support for having nested at-rule inside style rule.
  - CSSOM
  - selector matching

* LayoutTests/fast/css/style-enumerate-properties-expected.txt:
* LayoutTests/fast/css/style-enumerate-properties.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/serialize-group-rules-with-decls.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/serialize-group-rules-with-decls.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/nesting-parsing-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/nesting-parsing.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/old-tests/css3-modsel-156.xml:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/old-tests/css3-modsel-156b.xml:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/old-tests/css3-modsel-156c.xml:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/Element-matches-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/Element-webkitMatchesSelector-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/ParentNode-querySelector-All-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/ParentNode-querySelector-All-xht-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/selectors.js:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::cssRules const):
* Source/WebCore/css/CSSComputedStyleDeclaration.h:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::simpleSelectorSpecificity):
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/CSSStyleDeclaration.h:
* Source/WebCore/css/CSSStyleDeclaration.idl:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::cssText const):
* Source/WebCore/css/PropertySetCSSStyleDeclaration.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::MutableStyleProperties::createEmpty):
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::averageSizeInBytes):
(WebCore::StyleRule::StyleRule):
(WebCore::StyleRule::create):
(WebCore::StyleRule::createForSplitting):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSelector):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::consumeRuleList):
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
(WebCore::CSSParserImpl::consumeDeclarationList):
(WebCore::CSSParserImpl::consumeStyleBlock):
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
(WebCore::CSSParserImpl::consumeStyleBlock):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::CSSSelectorParser::consumeSimpleSelector):
(WebCore::CSSSelectorParser::consumeNesting):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Canonical link: <a href="https://commits.webkit.org/258079@main">https://commits.webkit.org/258079@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b5c724d5313300ebb0c11640ef656b9225c528bd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100850 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10000 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33897 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110153 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170427 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104839 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10932 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/842 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93261 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107998 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/106633 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8262 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/91523 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34878 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/90170 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22919 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/77853 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/91336 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3690 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24437 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/87424 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/1220 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3709 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/802 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29233 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9820 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43938 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/90312 "Built successfully") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/5541 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/5487 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20219 "Passed tests") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2903 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/css/StyleRule.cpp:
```diff
@@ -199,30 +199,48 @@ Ref<CSSRule> StyleRuleBase::createCSSOMWrapper(CSSStyleSheet* parentSheet, CSSGr
 unsigned StyleRule::averageSizeInBytes()
 {
-    return sizeof(StyleRule) + sizeof(CSSSelector) + StyleProperties::averageSizeInBytes();
+    return sizeof(StyleRule) + sizeof(CSSSelector) + StyleProperties::averageSizeInBytes() + +sizeof(WeakPtr<StyleRule>) + sizeof(Vector<Ref<StyleRuleBase>>);
```
This looks wrong:

`StyleProperties::averageSizeInBytes() + +sizeof(WeakPtr<StyleRule>) + sizeof(`

I'm surprised it compiles!

brentfulgham commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -411,20 +411,22 @@ RefPtr<StyleRuleBase> CSSParserImpl::consumeAtRule(CSSParserTokenRange& range, A
         range.consumeComponentValue();
-
+    
+    
```
Nit: Extra whitespace.

brentfulgham commented on Source/WebCore/css/parser/CSSParserImpl.h:
```diff
@@ -125,7 +125,7 @@ class CSSParserImpl {
 
     // This function updates the range it's given.
-    RefPtr<StyleRuleBase> consumeQualifiedRule(CSSParserTokenRange&, AllowedRulesType);
+    RefPtr<StyleRuleBase> consumeQualifiedRule(CSSParserTokenRange&, AllowedRulesType, StyleRule* parentRule);
```
Would it make sense to have a default for parentRule?

`RefPtr<StyleRuleBase> consumeQualifiedRule(CSSParserTokenRange&, AllowedRulesType, StyleRule* parentRule = nullptr);`

mdubet replied:
I don't have a strong opinion on this, but default values are fairly rare in the WebKit codebase though.

mdubet replied:
Fixed to use a default value (null RefPtr).

brentfulgham commented on Source/WebCore/css/parser/CSSParserImpl.h:
```diff
@@ -144,9 +144,12 @@ class CSSParserImpl {
 
     RefPtr<StyleRuleKeyframe> consumeKeyframeStyleRule(CSSParserTokenRange prelude, CSSParserTokenRange block);
-    RefPtr<StyleRule> consumeStyleRule(CSSParserTokenRange prelude, CSSParserTokenRange block);
+    RefPtr<StyleRule> consumeStyleRule(CSSParserTokenRange prelude, CSSParserTokenRange block, StyleRule* parentRule);
```
Ditto: nullptr?

brentfulgham commented on Source/WebCore/css/parser/CSSParserImpl.h:
```diff
@@ -144,9 +144,12 @@ class CSSParserImpl {
+    // FIXME: We should return value for all those functions instead of using class member attributes.
+    void consumeDeclarationListOrStyleBlockHelper(CSSParserTokenRange, StyleRuleType, StyleRule* parentRule);
     void consumeDeclarationList(CSSParserTokenRange, StyleRuleType);
+    void consumeStyleBlock(CSSParserTokenRange, StyleRuleType, StyleRule* parentRule);
```
Ditto: nullptr?

brentfulgham commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -43,29 +43,36 @@ namespace WebCore {
+        if (m_context.cssNestingEnabled) {
+            if (m_isNestedSelector && range.peek().type() == SemicolonToken) {
+                range.consumeIncludingWhitespace();
+                return { };
```
Even if none of the above it true, we always fall through to the '`return { };`' below. Maybe this early return isn't that useful?

brentfulgham commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -988,20 +990,34 @@ static void observeSelectors(CSSParserObserverWrapper& wrapper, CSSParserTokenRa
 }
 
-RefPtr<StyleRule> CSSParserImpl::consumeStyleRule(CSSParserTokenRange prelude, CSSParserTokenRange block)
+RefPtr<StyleRule> CSSParserImpl::consumeStyleRule(CSSParserTokenRange prelude, CSSParserTokenRange block, StyleRule* parentRule)
```
Maybe we should we be passing a `std::optional<Ref<StyleRule>>` instead?

mdubet replied:
There is a small overhead by using a ref-counted pointer instead of raw pointer. In this kind of nested calls, we know for sure that the pointer will exists at least until we come back to the caller. But maybe it's safer/cleaner to use a `RefPtr` anyway.

brentfulgham commented on Source/WebCore/css/StyleRule.cpp:
```diff
@@ -199,30 +199,48 @@ Ref<CSSRule> StyleRuleBase::createCSSOMWrapper(CSSStyleSheet* parentSheet, CSSGr
 StyleRule::~StyleRule() = default;
 
-Ref<StyleRule> StyleRule::create(Ref<StyleProperties>&& properties, bool hasDocumentSecurityOrigin, CSSSelectorList&& selectors)
+void StyleRule::setProperties(Ref<StyleProperties> properties) { m_properties = properties; }
```
We don't usually define methods in implementation files in one line like this. Could this be in the header file instead? Otherwise we should split it into three lines so it's easier to debug.

mdubet replied:
Yes the goal was to avoid including `StyleProperties.h` in the header but indeed it's not following our guidelines. Fixed.

brentfulgham commented on Source/WebCore/css/parser/CSSSelectorParser.h:
```diff
@@ -98,6 +101,6 @@ class CSSSelectorParser {
 
 
-std::optional<CSSSelectorList> parseCSSSelector(CSSParserTokenRange, const CSSParserContext&, StyleSheetContents*);
+std::optional<CSSSelectorList> parseCSSSelector(CSSParserTokenRange, const CSSParserContext&, StyleSheetContents*, bool);
```
You actually should give this `bool` a name -- it's hard to know what it's for!

mdubet replied:
Yes, changed to use a proper `enum`

brentfulgham commented on Source/WebCore/css/StyleRule.cpp:
```diff
@@ -199,30 +199,48 @@ Ref<CSSRule> StyleRuleBase::createCSSOMWrapper(CSSStyleSheet* parentSheet, CSSGr
 }
 
-StyleRule::StyleRule(Ref<StyleProperties>&& properties, bool hasDocumentSecurityOrigin, CSSSelectorList&& selectors)
+StyleRule::StyleRule(Ref<StyleProperties>&& properties, bool hasDocumentSecurityOrigin, CSSSelectorList&& selectors, WeakPtr<StyleRule> parentRule, Vector<Ref<StyleRule>> nestedRules)
```
Could we make `Vector<Ref<StyleRule>> nestedRules` a move operation instead of a copy? Or pass as a const reference so we only copy on assignment to the member?

brentfulgham commented on Source/WebCore/css/StyleRule.cpp:
```diff
@@ -199,30 +199,48 @@ Ref<CSSRule> StyleRuleBase::createCSSOMWrapper(CSSStyleSheet* parentSheet, CSSGr
     , m_properties(WTFMove(properties))
     , m_selectorList(WTFMove(selectors))
+    , m_parentRule(parentRule)
+    , m_nestedRules(nestedRules)
```
Could this be `m_nestedRules(WTFMove(nestedRules))`?

anttijk commented on Source/WebCore/css/StyleRule.h:
```diff
@@ -127,16 +128,26 @@ class StyleRule final : public StyleRuleBase {
     CSSSelectorList m_selectorList;
 
+    // CSS Nesting
+    WeakPtr<StyleRule> m_parentRule;
```
I don't think we necessarily need a parent rule pointer. The patch doesn't use it so let's not introduce it here. If we build full selectors up in the RuleSetBuilder we will always have the parent available.

anttijk commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -43,29 +43,36 @@ namespace WebCore {
 static bool consumeANPlusB(CSSParserTokenRange&, std::pair<int, int>&);
 
-std::optional<CSSSelectorList> parseCSSSelector(CSSParserTokenRange range, const CSSParserContext& context, StyleSheetContents* styleSheet)
+std::optional<CSSSelectorList> parseCSSSelector(CSSParserTokenRange range, const CSSParserContext& context, StyleSheetContents* styleSheet, bool isNestedSelector)
```
Please use an enum class instead of a magic boolean.

mdubet committed:
CSS nesting: parsing of style rule inside style rule.
https://bugs.webkit.org/show_bug.cgi?id=249263
rdar://103146832

Reviewed by Antti Koivisto.

This implements experimental support to parse nested style rule inside style rule.
This is based on the css-nesting-1 editor draft as of today,
which describes the so-called "option 3" (non letter start) syntax.

https://github.com/w3c/csswg-drafts/blob/main/css-nesting-1/proposals.md

This is missing:
  - any support for having nested at-rule inside style rule.
  - CSSOM
  - selector matching

* LayoutTests/fast/css/style-enumerate-properties-expected.txt:
* LayoutTests/fast/css/style-enumerate-properties.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/cssom-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/parsing.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/serialize-group-rules-with-decls.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/serialize-group-rules-with-decls.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/nesting-parsing-expected.txt: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/nesting-parsing.html: Removed.
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/old-tests/css3-modsel-156.xml:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/old-tests/css3-modsel-156b.xml:
* LayoutTests/imported/w3c/web-platform-tests/css/selectors/old-tests/css3-modsel-156c.xml:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/Element-matches-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/Element-webkitMatchesSelector-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/ParentNode-querySelector-All-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/ParentNode-querySelector-All-xht-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/dom/nodes/selectors.js:
* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::CSSComputedStyleDeclaration::cssRules const):
* Source/WebCore/css/CSSComputedStyleDeclaration.h:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::simpleSelectorSpecificity):
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/CSSStyleDeclaration.h:
* Source/WebCore/css/CSSStyleDeclaration.idl:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::cssText const):
* Source/WebCore/css/PropertySetCSSStyleDeclaration.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::MutableStyleProperties::createEmpty):
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/StyleRule.cpp:
(WebCore::StyleRule::averageSizeInBytes):
(WebCore::StyleRule::StyleRule):
(WebCore::StyleRule::create):
(WebCore::StyleRule::createForSplitting):
* Source/WebCore/css/StyleRule.h:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSelector):
* Source/WebCore/css/parser/CSSParserContext.cpp:
(WebCore::operator==):
(WebCore::add):
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::parseRule):
(WebCore::CSSParserImpl::consumeRuleList):
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
(WebCore::CSSParserImpl::consumeDeclarationList):
(WebCore::CSSParserImpl::consumeStyleBlock):
* Source/WebCore/css/parser/CSSParserImpl.h:
(WebCore::CSSParserImpl::consumeQualifiedRule):
(WebCore::CSSParserImpl::consumeStyleRule):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
(WebCore::CSSParserImpl::consumeStyleBlock):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::CSSSelectorParser::consumeSimpleSelector):
(WebCore::CSSSelectorParser::consumeNesting):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Canonical link: https://commits.webkit.org/258079@main
